File: iOS/TUIKit/TUICallKit/TUICallKit/UI/TUICallingAction.m
//
//  TUICallingAction.m
//  TUICalling
//
//  Created by noah on 2022/5/17.
//  Copyright © 2022 Tencent. All rights reserved
//

#import "TUICallingAction.h"
#import <TUICore/TUICore.h>
#import <TUICore/TUIDefine.h>
#import "TUICallEngineHeader.h"
#import "TUICallingStatusManager.h"
#import "TUICommonModel.h"
#import "TUICallKitOfflinePushInfoConfig.h"
#import "TUICallingUserManager.h"
#import "TUICallKitConstants.h"

@implementation TUICallingAction

+ (void)accept {
    [[TUICallEngine createInstance] accept:^{
        [TUICallingStatusManager shareInstance].callStatus = TUICallStatusAccept;
    } fail:^(int code, NSString *errMsg) {
        [TUICallingStatusManager shareInstance].callStatus = TUICallStatusNone;
    }];
}

+ (void)reject {
    [[TUICallEngine createInstance] reject:^{
        [TUICallingStatusManager shareInstance].callStatus = TUICallStatusNone;
    } fail:^(int code, NSString *errMsg) {
        [TUICallingStatusManager shareInstance].callStatus = TUICallStatusNone;
    }];
}

+ (void)hangup {
    [[TUICallEngine createInstance] hangup:^{
        [TUICallingStatusManager shareInstance].callStatus = TUICallStatusNone;
    } fail:^(int code, NSString *errMsg) {
        [TUICallingStatusManager shareInstance].callStatus = TUICallStatusNone;
    }];
}

+ (void)switchToAudioCall {
    [[TUICallEngine createInstance] switchCallMediaType:TUICallMediaTypeAudio];
}

+ (void)openCamera:(TUICamera)camera videoView:(TUIVideoView *)videoView {
    [[TUICallEngine createInstance] openCamera:camera videoView:videoView succ:^{
        [TUICallingStatusManager shareInstance].isCloseCamera = NO;
    } fail:^(int code, NSString *errMsg) {
    }];
}

+ (void)closeCamera {
    [[TUICallEngine createInstance] closeCamera];
    [TUICallingStatusManager shareInstance].isCloseCamera = YES;
}

+ (void)switchCamera{
    TUICamera callingCamera = TUICameraFront;
    
    if ([TUICallingStatusManager shareInstance].camera == TUICameraFront) {
        callingCamera = TUICameraBack;
    }
    
    [[TUICallEngine createInstance] switchCamera:callingCamera];
    [TUICallingStatusManager shareInstance].camera = callingCamera;
}

+ (void)openMicrophone {
    if ([TUICallingStatusManager shareInstance].callStatus != TUICallStatusAccept) {
        [TUICallingStatusManager shareInstance].isMicMute = YES;
        return;
    }
    [[TUICallEngine createInstance] openMicrophone:^{
        [TUICallingStatusManager shareInstance].isMicMute = YES;
        [TUICore notifyEvent:TUICore_TUICallKitVoIPExtensionNotify
                      subKey:TUICore_TUICore_TUICallKitVoIPExtensionNotify_OpenMicrophoneSubKey
                      object:nil param:nil];
    } fail:^(int code, NSString *errMsg) {
    }];
}

+ (void)closeMicrophone {
    if ([TUICallingStatusManager shareInstance].callStatus != TUICallStatusAccept) {
        [TUICallingStatusManager shareInstance].isMicMute = NO;
        return;
    }
    [[TUICallEngine createInstance] closeMicrophone];
    [TUICore notifyEvent:TUICore_TUICallKitVoIPExtensionNotify
                  subKey:TUICore_TUICore_TUICallKitVoIPExtensionNotify_CloseMicrophoneSubKey
                  object:nil param:nil];
    [TUICallingStatusManager shareInstance].isMicMute = NO;
}

+ (void)selectAudioPlaybackDevice {
    TUIAudioPlaybackDevice device = TUIAudioPlaybackDeviceSpeakerphone;
    
    if ([TUICallingStatusManager shareInstance].audioPlaybackDevice == TUIAudioPlaybackDeviceSpeakerphone) {
        device = TUIAudioPlaybackDeviceEarpiece;
    }
    
    [[TUICallEngine createInstance] selectAudioPlaybackDevice:device];
    [TUICallingStatusManager shareInstance].audioPlaybackDevice = device;
}

+ (void)inviteUser:(NSArray<TUIUserModel *> *)userList succ:(void(^)(NSArray *userIDs))succ fail:(TUICallFail)fail {
    NSMutableArray *userIdList = [NSMutableArray array];
    for (TUIUserModel *userModel in userList) {
        if (userModel.userId) {
            [userIdList addObject:userModel.userId];
        }
    }
    if (userIdList.count <= 0) {
        if (fail) {
            fail(ERROR_PARAM_INVALID, @"inviteUser invalid userList");
        }
        return;
    }
    if ((userIdList.count + [TUICallingUserManager allUserIdList].count) > 9) {
        if (fail) {
            fail(ERROR_PARAM_INVALID, @"TUICalling currently supports call with up to 9 people.");
        }
        return;
    }
    TUIOfflinePushInfo *offlinePushInfo = [TUICallKitOfflinePushInfoConfig createOfflinePushInfo];
    TUICallParams *callParams = [TUICallParams new];
    callParams.offlinePushInfo = offlinePushInfo;
    callParams.timeout = TUI_CALLKIT_SIGNALING_MAX_TIME;
    [[TUICallEngine createInstance] inviteUser:[userIdList copy] params:callParams succ:succ fail:fail];
}

+ (void)startRemoteView:(NSString *)userId
              videoView:(TUIVideoView *)videoView
              onPlaying:(void(^)(NSString *userId))onPlaying
              onLoading:(void(^)(NSString *userId))onLoading
                onError:(void(^)(NSString *userId, int code, NSString *errMsg))onError {
    [[TUICallEngine createInstance] startRemoteView:userId videoView:videoView onPlaying:onPlaying onLoading:onLoading onError:onError];
}

+ (void)stopRemoteView:(NSString *)userId {
    [[TUICallEngine createInstance] stopRemoteView:userId];
}

@end

File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/base/TUICallingAction.java
package com.tencent.qcloud.tuikit.tuicallkit.base;

import android.content.Context;

import com.tencent.qcloud.tuikit.TUICommonDefine;
import com.tencent.qcloud.tuikit.TUIVideoView;
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine;
import com.tencent.qcloud.tuikit.tuicallengine.TUICallEngine;
import com.tencent.qcloud.tuikit.tuicallkit.config.OfflinePushInfoConfig;

import java.util.List;

public class TUICallingAction {
    private final Context mContext;

    public TUICallingAction(Context context) {
        mContext = context;
    }

    public void inviteUser(List<String> userIdList, TUICommonDefine.ValueCallback callback) {
        TUICallDefine.CallParams params = new TUICallDefine.CallParams();
        params.offlinePushInfo = OfflinePushInfoConfig.createOfflinePushInfo(mContext);
        params.timeout = Constants.SIGNALING_MAX_TIME;
        TUICallEngine.createInstance(mContext).inviteUser(userIdList, params, callback);
    }

    public void accept(TUICommonDefine.Callback callback) {
        TUICallEngine.createInstance(mContext).accept(new TUICommonDefine.Callback() {
            @Override
            public void onSuccess() {
                TUICallingStatusManager.sharedInstance(mContext).updateCallStatus(TUICallDefine.Status.Accept);
                if (callback != null) {
                    callback.onSuccess();
                }
            }

            @Override
            public void onError(int errCode, String errMsg) {
                TUICallingStatusManager.sharedInstance(mContext).updateCallStatus(TUICallDefine.Status.None);
                if (callback != null) {
                    callback.onError(errCode, errMsg);
                }
            }
        });
    }

    public void reject(TUICommonDefine.Callback callback) {
        TUICallEngine.createInstance(mContext).reject(new TUICommonDefine.Callback() {
            @Override
            public void onSuccess() {
                TUICallingStatusManager.sharedInstance(mContext).updateCallStatus(TUICallDefine.Status.None);
                if (callback != null) {
                    callback.onSuccess();
                }
            }

            @Override
            public void onError(int errCode, String errMsg) {
                TUICallingStatusManager.sharedInstance(mContext).updateCallStatus(TUICallDefine.Status.None);
                if (callback != null) {
                    callback.onError(errCode, errMsg);
                }
            }
        });
    }

    public void hangup(TUICommonDefine.Callback callback) {
        TUICallEngine.createInstance(mContext).hangup(new TUICommonDefine.Callback() {
            @Override
            public void onSuccess() {
                TUICallingStatusManager.sharedInstance(mContext).updateCallStatus(TUICallDefine.Status.None);
                if (callback != null) {
                    callback.onSuccess();
                }
            }

            @Override
            public void onError(int errCode, String errMsg) {
                TUICallingStatusManager.sharedInstance(mContext).updateCallStatus(TUICallDefine.Status.None);
                if (callback != null) {
                    callback.onError(errCode, errMsg);
                }
            }
        });
    }

    public void switchCallMediaType(TUICallDefine.MediaType callMediaType) {
        TUICallEngine.createInstance(mContext).switchCallMediaType(callMediaType);
    }

    public void openCamera(TUICommonDefine.Camera camera, TUIVideoView videoView, TUICommonDefine.Callback callback) {
        TUICallEngine.createInstance(mContext).openCamera(camera, videoView, new TUICommonDefine.Callback() {
            @Override
            public void onSuccess() {
                TUICallDefine.Status status = TUICallingStatusManager.sharedInstance(mContext).getCallStatus();
                if (!TUICallDefine.Status.None.equals(status)) {
                    TUICommonDefine.Camera camera = TUICallingStatusManager.sharedInstance(mContext).getFrontCamera();
                    TUICallingStatusManager.sharedInstance(mContext).updateCameraOpenStatus(true, camera);
                    TUICallingStatusManager.sharedInstance(mContext).updateFrontCameraStatus(camera);
                }

                if (callback != null) {
                    callback.onSuccess();
                }
            }

            @Override
            public void onError(int errCode, String errMsg) {
                if (callback != null) {
                    callback.onError(errCode, errMsg);
                }
            }
        });
    }

    public void closeCamera() {
        TUICallEngine.createInstance(mContext).closeCamera();
        TUICommonDefine.Camera frontCamera = TUICallingStatusManager.sharedInstance(mContext).getFrontCamera();
        TUICallingStatusManager.sharedInstance(mContext).updateCameraOpenStatus(false, frontCamera);
    }

    public void switchCamera(TUICommonDefine.Camera camera) {
        TUICallEngine.createInstance(mContext).switchCamera(camera);
        TUICallingStatusManager.sharedInstance(mContext).updateFrontCameraStatus(camera);
    }

    public void openMicrophone(TUICommonDefine.Callback callback) {
        TUICallEngine.createInstance(mContext).openMicrophone(callback);
        TUICallingStatusManager.sharedInstance(mContext).updateMicMuteStatus(false);
    }

    public void closeMicrophone() {
        TUICallEngine.createInstance(mContext).closeMicrophone();
        TUICallingStatusManager.sharedInstance(mContext).updateMicMuteStatus(true);
    }

    public void selectAudioPlaybackDevice(TUICommonDefine.AudioPlaybackDevice device) {
        TUICallEngine.createInstance(mContext).selectAudioPlaybackDevice(device);
        TUICallingStatusManager.sharedInstance(mContext).updateAudioPlaybackDevice(device);
    }

    public void startRemoteView(String userId, TUIVideoView videoView, TUICommonDefine.PlayCallback callback) {
        TUICallEngine.createInstance(mContext).startRemoteView(userId, videoView, callback);
    }

    public void stopRemoteView(String userId) {
        TUICallEngine.createInstance(mContext).stopRemoteView(userId);
    }
}
